### PR TITLE
Auto batch smart reshape (relies on the dim tracking)

### PR DESCRIPTION
**Pull request by myshevts**

### CODE REVIEW

File: src/plugins/auto_batch/auto_batch.cpp [hunk 1]
```diff
@@ -6,13 +6,17 @@
...
 #include <ie_icore.hpp>
 #include <ie_ngraph_utils.hpp>
 #include <ie_performance_hints.hpp>
 #include <iostream>
 #include <map>
 #include <memory>
+#include <openvino/pass/manager.hpp>
 #include <string>
+#include <transformations/common_optimizations/dimension_tracking.hpp>
+#include <transformations/init_node_info.hpp>
```
Comment (ilya-lavrenov) [inline src/plugins/auto_batch/auto_batch.cpp:19]: use `""` for our headers.

File: src/plugins/auto_batch/auto_batch.cpp [hunk 2]
```diff
@@ -32,11 +36,9 @@ Blob::Ptr create_shared_blob_on_top_of_batched_blob(Blob::Ptr batched_blob, size
...
     auto layout = batched_blob->getTensorDesc().getLayout();
     SizeVector dims = batched_blob->getTensorDesc().getDims();
-    // the below code is a placeholder for the WIP (22.1) functionality
-    // that will check the reshaping by the batch is robust (CVS-51744)
-    if (layout == InferenceEngine::Layout::NC || layout == InferenceEngine::Layout::NCDHW ||
-        layout == InferenceEngine::Layout::NCHW || layout == InferenceEngine::Layout::NHWC ||
-        layout == InferenceEngine::Layout::NDHWC) {
+    std::stringstream str;
+    str << layout;
+    if (str.str().find("N") == 0) {
```
Comment (ilya-lavrenov) [inline src/plugins/auto_batch/auto_batch.cpp:41]: use `ov::layout::has_batch`

Comment (myshevts) [reply src/plugins/auto_batch/auto_batch.cpp:41]: tis wouldn't work when the code is called from the old api (e.g. legacy apps)

File: src/plugins/auto_batch/auto_batch.cpp [hunk 3]
```diff
@@ -670,64 +677,70 @@ InferenceEngine::IExecutableNetworkInternal::Ptr AutoBatchInferencePlugin::LoadN
...
-                layout == InferenceEngine::Layout::NDHWC) {
-                if (1 != item.second->getTensorDesc().getDims()[0])  // do not reshape/re-batch batched networks
+        // input(s) should have the batch dim as the first dim or none (current limitation of the auto-batching impl)
+        const auto& params = function->get_parameters();
+        for (size_t input_id = 0; input_id < params.size(); input_id++) {
+            const auto& input = params[input_id];
+            const auto& shape = input->get_partial_shape();
+            // currently no plugin support batched execution for dynamic networks
+            if (shape.is_dynamic())
+                IE_THROW(NotImplemented) << "Auto-batching does not reshape/re-batch originally batched networks!";
```
Comment (jane-intel) [inline src/plugins/auto_batch/auto_batch.cpp:724]: incorrect message. it's better to mention dynamism

File: src/plugins/auto_batch/auto_batch.cpp [hunk 4]
```diff
@@ -670,64 +677,70 @@ InferenceEngine::IExecutableNetworkInternal::Ptr AutoBatchInferencePlugin::LoadN
...
+            // check the batch dim: either 0th (and the original batch size of 1) or none
+            if (ov::DimensionTracker::get_label(shape[0])) {
```
Comment (jane-intel) [inline src/plugins/auto_batch/auto_batch.cpp:726]: check if shape.size() > 0

**Review by jane-intel — APPROVED** [review]
Pretty limited, but still great usage of FindBatch

**Pull request merged**